File: python/taflow/indicators/cup_and_handle.py
"""Python adapter for native cup-and-handle detection."""

from typing import Any

import numpy as np

from .._native import CupAndHandle as _Native
from .._series import as_float64_series


class CupAndHandle:
    """Detect causal cup-and-handle and inverse formations.

    A native 5% non-repainting swing tracker examines four confirmed pivots.
    Matching rims within 3% followed by a shallow handle emit ``1.0`` for a
    bullish cup or ``-1.0`` for an inverse cup; every other bar emits ``0.0``.
    Although open and close are accepted for a uniform OHLC interface, geometry
    uses high and low. The definition maps to Wickra ``CupAndHandle``.

    Args:
        open: Required opening-price history.
        high: Required high-price history.
        low: Required low-price history.
        close: Required closing-price history.

    Raises:
        ValueError: If the four OHLC histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(
        self, open: float, high: float, low: float, close: float
    ) -> "CupAndHandle":
        """Append one OHLC bar and return this adapter."""
        self._state.append(float(open), float(high), float(low), float(close))
        return self

    def extend(self, open: Any, high: Any, low: Any, close: Any) -> "CupAndHandle":
        """Append aligned OHLC histories after validating their lengths."""
        arrays = tuple(as_float64_series(item) for item in (open, high, low, close))
        if len({len(item) for item in arrays}) != 1:
            raise ValueError("OHLC inputs must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest pattern signal, or ``None`` before the first bar."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned ``-1``, ``0``, and ``1`` signal history."""
        return self._state.compute()

    def reset(self) -> "CupAndHandle":
        """Clear confirmed pivots and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/cup_and_handle.rs
use crate::error::TaResult;

use super::pattern_swing::{approximately_equal, SwingTracker, LEVEL_TOLERANCE, SWING_THRESHOLD};

/// Causal cup-and-handle and inverse-cup-and-handle detector.
#[derive(Debug, Clone)]
pub struct CupAndHandle {
    swing: SwingTracker,
    count: usize,
    value: Option<f64>,
}

impl CupAndHandle {
    /// Create a detector using Wickra's fixed 5% swing threshold.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            swing: SwingTracker::new(SWING_THRESHOLD, 4),
            count: 0,
            value: None,
        })
    }

    /// Append one OHLC bar and return `1`, `-1`, or `0` from the first bar.
    pub fn append(&mut self, _open: f64, high: f64, low: f64, _close: f64) -> Option<f64> {
        self.count += 1;
        self.value = Some(0.0);
        if !self.swing.append(high, low) {
            return self.value;
        }
        let pivots = self.swing.pivots();
        if pivots.len() < 4 {
            return self.value;
        }

        let length = pivots.len();
        let left_rim = pivots[length - 4];
        let extreme = pivots[length - 3];
        let right_rim = pivots[length - 2];
        let handle = pivots[length - 1];
        let rims_match = approximately_equal(left_rim.price, right_rim.price, LEVEL_TOLERANCE);
        self.value = if handle.direction < 0.0
            && rims_match
            && handle.price > extreme.price
            && handle.price < right_rim.price
        {
            Some(1.0)
        } else if handle.direction > 0.0
            && rims_match
            && handle.price < extreme.price
            && handle.price > right_rim.price
        {
            Some(-1.0)
        } else {
            Some(0.0)
        };
        self.value
    }

    /// Return the signal produced for the latest bar.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Return the number of processed bars.
    pub fn len(&self) -> usize {
        self.count
    }

    /// Return whether no bars have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }

    /// Clear pivots and restore fresh-state behaviour.
    pub fn reset(&mut self) {
        self.swing.reset();
        self.count = 0;
        self.value = None;
    }
}
